### DISCUSSION TITLE
valuesFrom: meaning of priority in 'These values always have a lower priority than the values inlined in the HelmRelease via the spec.values parameter.'

### DISCUSSION THREAD

**Discussion by cer** [fluxcd/flux2#4164] [Q&A]
https://fluxcd.io/flux/guides/helmreleases/#refer-to-values-in-configmap-and-secret-resources

> It is possible to define a list of ConfigMap and Secret resources from which to take values. The values are merged in the **order given**, with the later values overwriting earlier. These values always have a **lower priority** than the values inlined in the HelmRelease via the spec.values parameter.

What does 'lower priority' mean? `values` overrides what's specified by `valuesFrom` or vice versa?

Perhaps using **order** rather than **priority** would be clearer?

**Comment by makkes** [comment]
> What does 'lower priority' mean? values overrides what's specified by valuesFrom or vice versa?

`.spec.values` override `.spec.valuesFrom`.

> Perhaps using order rather than priority would be clearer?

I agree that a list or even an example would definitely help here. Happy to review a PR improving this doc item.

**Reply by cer** [reply]
Thanks.

A follow up question: what do 'override' and 'merge' actually mean?

**Reply by makkes** [reply]
If a `.spec.valuesFrom` ConfigMap has this `.data`:

```yaml
foo:
  bar1: baz1
  bar2: baz2
```

and `.spec.values` has this:

```yaml
foo:
  bar2: baz3
```

then the result will be:

```yaml
foo:
  bar1: baz1
  bar2: baz3
```

This is the result of merging the two maps in the order `.spec.valuesFrom`, then `.spec.values`.

**Reply by cer** [reply]
Thanks. How are array elements that are themselves objects merged/overridden?

**Reply by makkes** [reply]
Array elements aren't deeply merged. So if you have `.spec.ValuesFrom` with a CM with this `.data`:

```yaml
foo:
- bar1: baz1
  bar2: baz2
```
and `.spec.values` has this:
```yaml
foo:
- bar2: baz3
```
then the result will be:
```yaml
foo:
- bar2: baz3
```
